application-mocca-calendar-api/src/main/java/org/xwiki/contrib/moccacalendar/internal/importJob/ImportJob.java

```diff
@@ -0,0 +1,266 @@
+{
+    /**
+     * Mocca calendar import job type.
+     */
+    public static final String JOB_TYPE = "moccacalendar.import";
+
+    private static final String NEW_EVENT_HOME = "WebHome";
+
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<String> documentReferenceResolver;
+
+    @Inject
+    private Provider<XWikiContext> wikiContextProvider;
+
+    @Inject
+    private FullCalendarManager fullCalendarManager;
+
+    @Inject
+    private HTMLConverter htmlConverter;
+
+    @Override
+    public String getType()
+    {
+        return JOB_TYPE;
+    }
+
+    @Override
+    public JobGroupPath getGroupPath()
+    {
+        List<String> path = new ArrayList<>();
+        path.add("moccacalendar");
+        path.add("import");
+        path.add(request.getParentRef());
+        return new JobGroupPath(path);
+    }
+
+    @Override
+    protected ImportJobStatus createNewStatus(ImportJobRequest request)
+    {
+        return new ImportJobStatus(JOB_TYPE, request, observationManager, loggerManager);
+    }
+
+    @Override
+    protected void runInternal()
+    {
+        try {
+            List<MoccaCalendarEvent> calendarEventsJson =
+                fullCalendarManager.getICalEventsFromFile(getICSFileContent(request.getFile()), null, null, true);
+            List<XWikiDocument> eventDocuments = new ArrayList<>();
+            this.progressManager.pushLevelProgress(calendarEventsJson.size() + 1, this);
+
+            for (MoccaCalendarEvent calendarEvent : calendarEventsJson) {
+                progressManager.startStep(this);
+                if (status.isCanceled()) {
+                    progressManager.endStep(this);
+                    break;
+                }
+
+                XWikiDocument eventDoc =
+                    getUniqueEventName(calendarEvent.getTitle().trim(), request.getParentRef(), eventDocuments);
+                createCalendarObjects(eventDoc, calendarEvent);
+                eventDocuments.add(eventDoc);
+
+                progressManager.endStep(this);
+                Thread.yield();
+            }
+            batchSave(eventDocuments);
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        } finally {
+            this.progressManager.popLevelProgress(this);
+        }
+    }
+
+    private void batchSave(List<XWikiDocument> eventDocuments) throws XWikiException
+    {
+        if (!status.isCanceled()) {
+            XWikiContext wikiContext = wikiContextProvider.get();
+            XWiki wiki = wikiContext.getWiki();
+            progressManager.startStep(this);
+            for (XWikiDocument calendarEvent : eventDocuments) {
+                wiki.saveDocument(calendarEvent, wikiContext);
+            }
+            progressManager.endStep(this);
+        }
+    }
+
+    private void createCalendarObjects(XWikiDocument eventDoc, MoccaCalendarEvent component) throws XWikiException
```

Comment:
All methods below this one I feel could be moved to another component, in order to respect the `single responsibility` principle. This way you would have this class that is simply about managing the job, and then another one that does all that hands-on processing of calendar events
Maybe a `CalendarEventImport`